### PR TITLE
NFT Mint As A Service

### PR DESCRIPTION
**Pull request by echohtp**
## Goals

Allow application developers to create NFTs with a single API call

App POSTS to `<MINT_SERVICE>/create`, possible payload

```
{
	destination_wallet: SolanaPubkey,
	...MetaplexMetadataJson 
}
```

## Concerns
on-chain collections and the nft association, Metaplex NFT Standard v1.1.0 no longer includes a `collections` attribute field.

### CODE REVIEW

File: src/graphql/mintNfts.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      description: 'Mint hash of newly minted NFT',
+    });
+  },
+});
+
+export const MintMetadata = inputObjectType({
+  name: 'MintMetadata',
+  description: 'The NFT metadata JSON',
+  definition (t) {
+    t.nonNull.string('metadataJson', {
```
Comment (kevinrodriguez-io) [inline src/graphql/mintNfts.ts:28]: This shouldn't be a raw stringified JSON as it makes the api a bit harder to use.

File: src/graphql/mintNfts.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      description: 'Base58 Encoded Box',
+    });
+    t.nonNull.string('nonce', {
+      description: 'Base58 Encoded nonce used for boxing the message',
+    });
+    t.nonNull.string('clientPublicKey', {
+      description: 'Base58 Encoded Client public key used to box the message',
+    });
+  },
+});
```
Comment (kevinrodriguez-io) [inline src/graphql/mintNfts.ts:49]: Remind me to remove this once merged to keep it on a single file

File: src/graphql/mintNfts.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    const connection = new Connection(clusterApiUrl('mainnet-beta'));
+    let uri,
+      nft,
+      wallet = null;
+
+    const clientSecret = msg;
+
+    // Get create wallet from the client secret
+    try {
+      wallet = Keypair.fromSecretKey(utf8Encode.encode(clientSecret));
```
Comment (kevinrodriguez-io) [inline src/graphql/mintNfts.ts:77]: I think this works if the keypair is base58, but usually they are number[] so the syntax is a bit different, here's how I do it:
```ts
const keyPairBytes = JSON.parse(
      decryptEncodedPayload(args.encryptedMessage),
    ) as number[];
    const keyPair = web3.Keypair.fromSecretKey(Uint8Array.from(keyPairBytes));
    ```

File: src/graphql/mintNfts.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        message: `Error creating wallet from client secret: ${e.message}`,
+      };
+    }
+
+    // Get a new metaplex object
+    const metaplex = new Metaplex(connection).use(keypairIdentity(wallet));
+
+    // Upload NFT Metadata, use metadata from api call, assume NFT Standard format
+    try {
+      uri = await metaplex.nfts().uploadMetadata(args.metadataJson);
```
Comment (kevinrodriguez-io) [inline src/graphql/mintNfts.ts:89]: We need a better data structure for this metadataJson field. Maybe we can specify it as open or something, please define a custom input object type for this

**Review by kevinrodriguez-io — CHANGES_REQUESTED** [review]

File: src/graphql/mintNfts.ts (hunk 4 above)
Comment (echohtp) [reply src/graphql/mintNfts.ts:89]: agreed 👍🏽

File: src/graphql/mintNfts.ts (hunk 2 above)
Comment (echohtp) [reply src/graphql/mintNfts.ts:49]: I removed it, build was successful. It was initially useful to refer to the object but is redundant.

**Review by kevinrodriguez-io — APPROVED** [review]
LOOKS ABSOLUTELY GOOD.
Which makes me thing we should add a little nodejs cli client to test this things.

What do you think?

**Comment by echohtp** [comment]
> es me thing we should add a little nodejs cli client to test this things.

I would love to add some testing to this - lets talk about it

**Pull request merged**